File: DALNEW/GenericDAO.vb
Imports System.Linq
Imports System.Linq.Expressions

Public Class GenericDAO(Of t As Class)
    Implements IDisposable

    Private _dbSet As System.Data.Entity.DbSet(Of t)
    Private _context As testdbEntities
    Public Property Context() As testdbEntities
        Get
            Return _context
        End Get
        Set(ByVal value As testdbEntities)
            If (value IsNot Nothing) Then
                _context = value
                _dbSet = _context.Set(Of t)()
            End If
        End Set
    End Property

    Public Sub Dispose() Implements IDisposable.Dispose
        If (_context IsNot Nothing) Then
            _context.Dispose()
            _context = Nothing
        End If
    End Sub

    Public Function GetNormal() As IQueryable(Of t)
        Return _dbSet
    End Function

    Public Function GetNoTracking() As IQueryable(Of t)
        Return _dbSet.AsNoTracking()
    End Function
    Public Function GetNormal(filter As Expression(Of Func(Of t, Boolean))) As IQueryable(Of t)
        Dim query As IQueryable(Of t) = _dbSet
        If filter IsNot Nothing Then
            query = query.Where(filter)
        End If
        Return query
    End Function

    Public Function GetById(id As Integer) As t
        Return _dbSet.Find(id)
    End Function
    Public Function DeleteObject(t)
        Return _dbSet.Remove(t)
    End Function
    Public Function DeleteObject(id As Integer)
        Dim entity = _dbSet.Find(id)

        If (entity IsNot Nothing) Then

            Return _dbSet.Remove(entity)
        End If
        Return Nothing
    End Function


    Public Function GetNew() As t
        Dim retval = _dbSet.Create()
        _dbSet.Add(retval)
        Return retval
    End Function

End Class

File: DALCore/GenericDAO.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;
using DALCore.Models;
using Microsoft.EntityFrameworkCore;

namespace DALCore
{
    public class GenericDAO<T> : IDisposable where T : class
    {
        private DbSet<T> _dbSet;
        private cpmRunningContext _Context;
        public cpmRunningContext Context
        {
            get
            {
                return _Context;
            }
            set
            {
                if ((value != null))
                {
                    _Context = value;
                    _dbSet = _Context.Set<T>();
                }
            }
        }

        public void Dispose()
        {
            if ((_Context != null))
            {
                _Context.Dispose();
                _Context = null;
            }
        }

        public IQueryable<T> GetNormal()
        {
            return _dbSet;
        }

        public IQueryable<T> GetNoTracking()
        {
            return _dbSet.AsNoTracking();
        }
        public IQueryable<T> GetNormal(Expression<Func<T, bool>> filter)
        {
            IQueryable<T> query = _dbSet;
            if (filter != null)
                query = query.Where(filter);
            return query;
        }

        public T GetById(int id)
        {
            return _dbSet.Find(id);
        }
        public void DeleteObject(T t)
        {
            _dbSet.Remove(t);
        }
        public void DeleteObject(int id)
        {
            var entity = _dbSet.Find(id);

            if ((entity != null))
                _dbSet.Remove(entity);
            return;
        }


        public T GetNew()
        {
            var retval = Activator.CreateInstance<T>(); // maakt een gewone instance aan
            //var retval = _dbSet.CreateProxy();
            _dbSet.Add(retval);
            return retval;
        }
    }

}
